nixos/modules/security/apparmor.nix

```diff
@@ -1,59 +1,190 @@
+      lib.mapAttrsToList (name: path: {inherit name path;}) cfg.includes
+    );
+    environment.etc."apparmor/parser.conf".text = ''
+      ${if cfg.enableCache then "write-cache" else "skip-cache"}
+      cache-loc /var/cache/apparmor
+      Include /etc/apparmor.d
+    '' +
+    lib.concatMapStrings (p: "Include ${p}/etc/apparmor.d\n") cfg.packages;
+    environment.etc."apparmor/logprof.conf".text = ''
+      [settings]
+        profiledir = /etc/apparmor.d
+        inactive_profiledir = ${pkgs.apparmor-profiles}/share/apparmor/extra-profiles
+        logfiles = /var/log/audit/audit.log /var/log/syslog /var/log/messages
+
+        parser = ${pkgs.apparmor-parser}/bin/apparmor_parser
+        ldd = ${pkgs.glibc.bin}/bin/ldd
+        logger = ${pkgs.utillinux}/bin/logger
+
+        # customize how file ownership permissions are presented
+        # 0 - off
+        # 1 - default of what ever mode the log reported
+        # 2 - force the new permissions to be user
+        # 3 - force all perms on the rule to be user
+        default_owner_prompt = 1
+
+        # custom directory locations to look for #includes
+        #
+        # each name should be a valid directory containing possible #include
+        # candidate files under the profile dir which by default is /etc/apparmor.d.
+        #
+        # So an entry of my-includes will allow /etc/apparmor.d/my-includes to
+        # be used by the yast UI and profiling tools as a source of #include
+        # files.
+        custom_includes =
+
+      [qualifiers]
+        ${pkgs.runtimeShell} = icnu
+        ${pkgs.bashInteractive}/bin/sh = icnu
+        ${pkgs.bashInteractive}/bin/bash = icnu
+    '' + head (match "^.*\\[qualifiers](.*)" # Drop the original [settings] section.
+                     (readFile "${pkgs.apparmor-utils}/etc/apparmor/logprof.conf"));
+
+    boot.kernelParams = [ "apparmor=1" "security=apparmor" ];
+
+    systemd.services.apparmor = {
+      after = [
+        "local-fs.target"
+        "systemd-journald-audit.socket"
+      ];
+      before = [ "sysinit.target" ];
+      wantedBy = [ "multi-user.target" ];
+      restartTriggers = [
+        etc."apparmor/parser.conf".source
+        etc."apparmor.d".source
+      ];
+      unitConfig = {
+        Description="Load AppArmor policies";
+        DefaultDependencies = "no";
+        ConditionSecurity = "apparmor";
+      };
+      # Reloading instead of restarting enables to load new AppArmor profiles
+      # without necessarily restarting all services which have Requires=apparmor.service
+      # It works by:
+      # - Adding or replacing into the kernel profiles enabled in cfg.policies
+      #   (because AppArmor can do that without stopping the processes already confined).
+      # - Removing from the kernel any profile whose name is not
+      #   one of the names within the content of the profiles in cfg.policies.
+      # - Killing the processes which are unconfined but now have a profile loaded
```

Comment:
How does this work? It seems a bit scary to me to kill processes.

Reply:
It seems scarier to me to have processes around actively circumventing the profiles, without that being obvious to the user

Reply:
I suppose killing the unconfined confinable processes by default poses a certain risk of causing people who casually enable AppArmor and get their processes killed unexpectedly to forsake AppArmor with an unnecessarily poor opinion of it.  As a compromise, I propose that `killUnconfinedConfinables` be disabled by default but the main `security.apparmor.enable` option's documentation recommend enabling it.

Reply:
Also asking upstream: @stevebeattie @cboltz This procedure seems a bit unorthodox, what is your recommendation here?

Reply:
It's indeed a bit unorthodox ;-)

I'm not familiar with NixOS, which makes recommending something a bit difficult. Please read the following lines with a grain of salt.

My recommendation not to use `aa-teardown` (and therefore not unloading any profiles) in my other comment is probably the most important part, because it will avoid that processes become accidently unconfined if someone runs `systemctl restart apparmor`.

That will however still leave the case of newly added profiles for already running processes, where killing/restarting the process is the only way to get them confined. Killing them by default is a secure way, but I'm not sure if the users will like it ;-)

Reply:
I think the biggest difference from a apparmor perspective between NixOS and other Linux distributions here is that new versions executables will get a new path in the nix store, while common Linux distributions will update executables in place.